Milestone 3 - Users & posts

macnick opened the pull request:
Here is what we implemented in this milestone

1. Create User Model
2. Create Post Model
3. Install Devise gem for user authentication
4. Log In and Sign Up pages for users to create an account and log in
5. Automatic log in after account creation
6. User page where he/she can create posts and view all the his/hers previous posts
7. Capability to delete a post.
8. Unit tests for User and Post models
9. Integration tests for user log in and sign up

Please review our code.

addod19 committed:
database configured

addod19 committed:
created devise user model, devise user view, and its migration

addod19 committed:
created users controller for our basic view actions

addod19 committed:
created users controller action with its associted view template

addod19 committed:
added user relationship to post

addod19 committed:
added boostrap gem, partials for header,default,falsh msg

addod19 committed:
configured permitted parameters in application controller and update the view

addod19 committed:
added signup helper

addod19 committed:
update sign out link with delete method and update routes and views to show page

addod19 committed:
added rubocop and stickler file

addod19 committed:
updated the fields in devise view

addod19 committed:
updated routes and header with logic if signed in

macnick committed:
Add posts migration

macnick committed:
Add controller and views for posts

addod19 committed:
added rspec rails to development test in gemfile

addod19 committed:
installed rspec

addod19 committed:
implemented unit test for user_spec

addod19 committed:
updated gem files to run rspec successfully

addod19 committed:
added icon to headers and custom styling to forms and some failing test

macnick committed:
Add hirb gem

macnick committed:
Nothing important

macnick committed:
Re-style the login and signup

macnick committed:
Require authentication, route to logged user's page

addod19 committed:
create post index view and it's controller action

macnick committed:
Show user past comments and add a button to delete a comment

macnick committed:
Add rspec test for User and Post models

addod19 committed:
user integration test but failing

macnick committed:
Fix Selenium::Webdriver error

addod19 committed:
Test pass now thanks to team work

macnick committed:
Fix the rspec errors

addod19 committed:
rubocop errors fix

Torres-ssf reviewed (changes requested):
### Changes required :smile: 
Hi there team :hand: Good job so far, but we have changes to make before moving on.

- [x] I notice that you added the linter files to your repo but we also need to give it the authorization for checking your repo files. Check out [this](https://github.com/microverseinc/linters-config/tree/master/ruby) link.
![Screenshot from 2020-01-06 21-14-16](https://user-images.githubusercontent.com/30977845/71858324-8b309c80-30c9-11ea-8b39-13658ef112bb.png)

- [x] The user's index controller needs to show all registered users. You don't need to create any buttons to add friends right now. Just make sure that the index method loads a view that shows all registered users' names. This will be improved in future milestones.
![Screenshot from 2020-01-08 10-04-44](https://user-images.githubusercontent.com/30977845/71980225-507c5080-31fe-11ea-8f8b-4daffab2e59f.png)

- [x] Good job implement the create post functionality :+1: . But as it said in the Odin project instructions, we need to display created posts with content and author.
![Screenshot from 2020-01-08 09-59-49](https://user-images.githubusercontent.com/30977845/71979999-cd5afa80-31fd-11ea-94ed-255b03bc167b.png)

- [x] Right now, one of your tests is falling
![Screenshot from 2020-01-08 10-20-01](https://user-images.githubusercontent.com/30977845/71981158-8c181a00-3200-11ea-82d9-118c31ac6190.png)

Please submit for code review again after making these changes.
Feel free to message me on slack if you have any questions
Happy coding :computer:

addod19 committed:
remved users link in header and re-positioned the links to the right

macnick committed:
Update README with optional install

macnick committed:
Merge branch 'Milestone-3' of https://github.com/macnick/facebook-clone into Milestone-3

macnick committed:
Remove uneeded comments from Gemfile

macnick committed:
Add the user name inside each post

macnick committed:
Change root route to users/index. View author on posts

macnick commented:
Hello, 
Thank you for the prompt review. It was very helpful.

We fixed the following: 
1. Gave authorization to stickler to check our repo. It created a pull request out of Milestone-3 and found no errors. We will wait for your ok to merge it with master.
2. Changed the  index controller to show all registered users with a fake button "Add friend" 
3. Modified posts index to include author for each post and of course the post text.
4. We forgot to write that for the first test to work you have to create a user with email "nick@test.com" and password "123456" first. The test, tests that an registered user can login. In a new database there is no registered user. As you see in the second test we create random users. 

Please review our changes.

addod19 committed:
fix link to create post in navbar link on profile

macnick committed:
Fix header menu alignment

macnick committed:
Merge header change to link

mosaaleb commented on .rubocop.yml:
```diff
@@ -0,0 +1,34 @@
+    - "Rakefile"
+  DisplayCopNames: true
+
+Metrics/LineLength:
```
- [x] There is no `Metrics/LineLength` rule. It's `Layout/LineLength`. Check it [here](https://rubocop.readthedocs.io/en/latest/cops_layout/#layoutlinelength)

mosaaleb commented on app/controllers/posts_controller.rb:
```diff
@@ -0,0 +1,29 @@
+  end
+
+  def create
+    @user = current_user
```
- [x] **[OPTIONAL]** `current_user` is already available in our views. We can remove this line and use `current_user` for the post creation.

addod19 replied:
removed line 10 and set @post = current_user.posts.build(post_params) as required by reviewer.

mosaaleb commented on app/controllers/posts_controller.rb:
```diff
@@ -0,0 +1,29 @@
+  private
+
+  def post_params
+    params.require(:post).permit(:post_text, :user_id)
```
- [x] **[OPTOINAL]** We don't need to permit the `user_id`. We already creating the post through association with `current_user`

addod19 replied:
removed user_id from post_params

mosaaleb commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,11 @@
+  end
+
+  def show
+    @user = User.find(current_user.id)
```
- [x] All show pages for all users are currently only displaying the current user. -- we need find the user from the id params `users/1`

addod19 replied:
Can you please create two users and two posts to see all the posts. Attached is the screenshot of the post index page
![Screen Shot 2020-01-09 at 11 02 04](https://user-images.githubusercontent.com/46388113/72053633-57b76300-32bf-11ea-97e4-f63c338ab156.png)

mosaaleb reviewed (changes requested):
### Well done @macnick @addod19 :clap:


- [x] Nice job adding stickler, but there are still stickler errors. Let's fix them. See my comment for this error.

- [x] I created a post with a user, and was able to delete the post with another user. -- We can display the delete button in our views only if the current user is the author if the post. -- Don't forget to wrap this logic (checking whether the current user is the author) in a helper method. [helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)

- [x] Failed test: 
![failed-test](https://user-images.githubusercontent.com/4246637/72018549-8f4aee80-3270-11ea-91b8-c31385058f6f.png)
The problem with this test is that you don't have a user with these credentials yet. 

I suggest that you go through the first 4 lessons in this course [TDD](https://thoughtbot.com/upcase/test-driven-rails)

Also we don't have integration test for adding posts. -- let's add it.

#### Please handle these requirements and submit a new code review.

##### Muhammad Ebeid

addod19 committed:
fix some comment from reviewer

macnick committed:
Delete button appear only on current user's posts

macnick committed:
Fix rubocop.yml

macnick committed:
Fix User can create post test

macnick committed:
Fix rubocop errors

macnick committed:
Update .rubocop.yml

macnick committed:
Change the rubocop.yml to Layout/LineLength

macnick committed:
Merge branch 'Milestone-3' of https://github.com/macnick/facebook-clone into Milestone-3

macnick commented:
Hello,
Thank you for the prompt review. It was very helpful.

We fixed the following:

1. Fixed stickler to work properly.
2. Fixed the delete button to appear only to the current user's posts.
3. Changed the logic for the test so first it creates the users and then test for login
4. Added integration test for creating post.
5. Also fixed the optional requirements.

Please review our changes.

mosaaleb reviewed (approved):
### Well done Team :clap:

#### All requirements are handled. All tests have passed.

#### You can merge :+1: 

##### Muhammad Ebeid

addod19 commented:
Thank you for helping us.

macnick committed:
Merge branch 'development' into Milestone-3
